shopfloor_packing/services/cluster_picking.py

```diff
@@ -6,10 +6,118 @@
 from odoo.addons.base_rest.components.service import to_int
 from odoo.addons.component.core import Component
 
+from .packaging import PackagingAction
+
 
 class ClusterPicking(Component):
+
     _inherit = "shopfloor.cluster.picking"
 
+    def _get_available_delivery_packaging(self, picking):
+        model = self.env["stock.package.type"]
+        carrier = picking.ship_carrier_id or picking.carrier_id
+        wizard_obj = self.env["choose.delivery.package"]
+        delivery_type = (
+            carrier.delivery_type
+            if carrier.delivery_type not in ("fixed", False)
+            else "none"
+        )
+        wizard = wizard_obj.with_context(
+            current_package_carrier_type=delivery_type
+        ).new({"picking_id": picking.id})
+        if not carrier:
+            return model.browse()
+        return model.search(
+            wizard.package_type_domain,
+            order="number_of_parcels,name",
+        )
+
+    def list_delivery_packaging(self, picking_batch_id, picking_id, selected_line_ids):
+        """List available delivery packaging for given picking.
+
+        Transitions:
+        * select_delivery_packaging: list available delivery packaging, the
+        user has to choose one to create the new package
+        * select_package: when no delivery packaging is available
+        """
+        picking = self.env["stock.picking"].browse(picking_id)
+        message = self._check_picking_status(picking)
+        if message:
+            return self._response_for_start(message=message)
+        selected_lines = self.env["stock.move.line"].browse(selected_line_ids).exists()
+        delivery_packaging = self._get_available_delivery_packaging(picking)
+        if not delivery_packaging:
+            return self._response_for_select_package(
+                picking,
+                selected_lines,
+                message=self.msg_store.no_delivery_packaging_available(),
+            )
+        response = self._check_allowed_qty_done(picking, selected_lines)
+        if response:
+            return response
+        return self._response_for_select_delivery_packaging(picking, delivery_packaging)
+
+    def scan_package_action(self, picking_id, selected_line_ids, barcode):
+        """Scan a package, a lot, a product or a package to handle a line
+
+        When a package is scanned (only delivery ones), if the package is known
+        as the destination package of one of the lines or is the source package
+        of a selected line, the package is set to be the destination package of
+        all the lines to pack.
+
+        When a product is scanned, it selects (set qty_done = reserved qty) or
+        deselects (set qty_done = 0) the move lines for this product. Only
+        products not tracked by lot can use this.
+
+        When a lot is scanned, it does the same as for the products but based
+        on the lot.
+
+        When a packaging type (one without related product) is scanned, a new
```

Comment:
model changed from v14, not related to product anymore
```suggestion
        When a package type is scanned, a new
```